📖 Update Required Configuration (#576)
kubernetes-sigs/cluster-api-provider-openstack: docs/configuration.md

<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**  *generated with [DocToc](https://github.com/thlorenz/doctoc)*

- [Required configuration](#required-configuration)
  - [Operating system image](#operating-system-image)
  - [Public network](#public-network)
  - [Floating IP](#floating-ip)
  - [SSH authorized key](#ssh-authorized-key)
  - [OpenStack credential](#openstack-credential)

>   - [Avaialability zone](#avaialability-zone)

Reviewer: ```suggestion
  - [Availability zone](#avaialability-zone)
```

  - [DNS server](#dns-server)
- [Optional Configuration](#optional-configuration)
  - [Network Filters](#network-filters)
  - [Multiple Networks](#multiple-networks)
  - [Subnet Filters](#subnet-filters)
  - [Tagging](#tagging)
  - [Metadata](#metadata)
  - [Boot From Volume](#boot-from-volume)
  - [Timeout settings](#timeout-settings)
  - [Custom pod network CIDR](#custom-pod-network-cidr)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

# Required configuration

The cluster configuration file can be generated by using [`clusterctl config cluster`](https://cluster-api.sigs.k8s.io/user/quick-start.html#generating-the-cluster-configuration) command.
This command actually uses [the template file](../templates/cluster-template.yaml) and replace the values surrounded by `${}` with environment variables. You have to set all required environment variables in advance. The following sections explain some more details about what should be configured.

Note: You can use [the template file](../templates/cluster-template.yaml) by manually replacing values.

## Operating system image

> We currently depend on an update version of cloud-init otherwise the operating system choice is yours. The kubeadm bootstrap provider we're using also depends on some pre-installed software like a controller-runtime, kubelet, kubeadm, etc.. . For an examples how to build such an image take a look at [image-builder (openstack)](https://image-builder.sigs.k8s.io/capi/providers/openstack.html).

Reviewer: ```suggestion
We currently depend on an up-to-date version of cloud-init otherwise the operating system choice is yours. The kubeadm bootstrap provider we're using also depends on some pre-installed software like a container runtime, kubelet, kubeadm, etc.. . For an examples how to build such an image take a look at [image-builder (openstack)](https://image-builder.sigs.k8s.io/capi/providers/openstack.html).
```

The image can be referenced by exposing it as an environment variable `OPENSTACK_IMAGE_NAME`.

## Public network

The public network id can be obtained by using command,

```bash
openstack network list --external
```
The ID must be expose as an environment variable `OPENSTACK_EXTERNAL_NETWORK_ID`

Note: If your openstack cluster does not already have a public network, you should contact your cloud service provider. We will not review how to troubleshoot this here.

## Floating IP

You have to be able to create a floating IP in your OpenStack. You can create one using,

```bash
openstack floating ip create <public network>
```
The IP must be exposed as an environment variable `OPENSTACK_CONTROLPLANE_IP`.

Note: Only user with admin role can create a floating IP with specific IP.

## SSH authorized key

The ssh public key is required. This key does not need to be created by OpenStack key pair.

The public key must be exposed as an environment variable `OPENSTACK_SSH_AUTHORIZED_KEY`.

> If you want to login to each machine by ssh, you have to configure security groups. If `spec.managedSecurityGroups` of `OpenStackCluster` set to true, two security groups will be created and added to the instances. One is `k8s-cluster-${NAMESPACE}-${CLUSTER_NAME}-secgroup-controlplane`, another is `k8s-cluster-${NAMESPACE}-${CLUSTER_NAME}-secgroup-worker`. These security group rules include the kubeadm's [Check required ports](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/#check-required-ports) so that each node can not be logged in through ssh by default. Please add existing security group allowing ssh port to OpenStackMachineTemplate spec. Here is an example:

Reviewer: ```suggestion
If you want to login to each machine by ssh, you have to configure security groups. If `spec.managedSecurityGroups` of `OpenStackCluster` set to true, two security groups will be created and added to the instances. One is `k8s-cluster-${NAMESPACE}-${CLUSTER_NAME}-secgroup-controlplane`, another is `k8s-cluster-${NAMESPACE}-${CLUSTER_NAME}-secgroup-worker`. These security group rules include the kubeadm's [Check required ports](https://kubernetes.io/docs/setup/production-environment/tools/kubeadm/install-kubeadm/#check-required-ports) so that each node can not be logged in through ssh by default. Please add pre-existing security group allowing ssh port to OpenStackMachineTemplate spec. Here is an example:
```

  template:
    spec:
      securityGroups:
      - name: allow-ssh
```

## OpenStack credential

The [env.rc](./env.rc) script sets the environment variables related to credentials.

```bash
source env.rc <path/to/clouds.yaml> <cloud>
```

The following variables are set.

| Variable | Meaning |
 :----- | :-------- 
| OPENSTACK_CLOUD | The cloud name which is used as second argument |
| OPENSTACK_CLOUD_YAML_B64 | The secret used by Cluster API Provider OpenStack accessing OpenStack |
| OPENSTACK_CLOUD_PROVIDER_CONF_B64 | The content of [cloud.conf](https://kubernetes.io/docs/concepts/cluster-administration/cloud-providers/#cloud-conf) which is used by OpenStack cloud provider |
| OPENSTACK_CLOUD_CACERT_B64 | (Optional) The content of your custom CA file which can be specified in your clouds.yaml by `ca-file` |

> ## Avaialability zone

Reviewer: ```suggestion
## Availability zone
```

The availability zone names must be exposed as an environment variable `OPENSTACK_FAILURE_DOMAIN`.

## DNS server

> The DNS servers must be exposed as an environment variable `OPENSTACK_DNS_NAMESERVERS`.

Reviewer 2: is it true? must expose?

Author: Yes, this is true!
```bash
(python) ubuntu@k8s:~/cluster-api-provider-openstack$ ./bin/clusterctl config cluster capi-quickstart --kubernetes-version v1.17.3 --control-plane-machine-count=3 --worker-machine-count=3 --infrastructure openstack:v0.3.0 > test.yaml
Error: failed to perform variable substitution: value for variables [OPENSTACK_DNS_NAMESERVERS] is not set. Please set the value using os environment variables or the clusterctl config file
```

Reviewer 2: oops... I mean, technically , not whether we are checking it 
e.g whether the DNS is really needed.. not remember clearly

Author: we can create subnet without DNS server. But kubeadm needs dns resolution. The VM image have to be created with DNS settings if `OPENSTACK_DNS_NAMESERVERS` is not given.

Reviewer 2: ok, works for me , thanks~
